src/gcd_ops.rs

```diff
@@ -0,0 +1,200 @@
+//! Operators for finding greatest common divisor.
+
+use crate::{
+    arch::word::Word,
+    buffer::Buffer,
+    div, gcd,
+    ibig::IBig,
+    memory::MemoryAllocation,
+    ops::{ExtendedGcd, Gcd},
+    ubig::{Repr::*, UBig},
+};
+
+impl Gcd<UBig> for UBig {
+    type Output = UBig;
+
+    #[inline]
+    fn gcd(self, rhs: UBig) -> UBig {
+        match (self.into_repr(), rhs.into_repr()) {
+            (Small(word0), Small(word1)) => UBig::gcd_word(word0, word1),
+            (Small(word0), Large(buffer1)) => UBig::gcd_large_word(&buffer1, word0),
+            (Large(buffer0), Small(word1)) => UBig::gcd_large_word(&buffer0, word1),
+            (Large(buffer0), Large(buffer1)) => UBig::gcd_large(buffer0, buffer1),
+        }
+    }
+}
+
+impl Gcd<&UBig> for UBig {
+    type Output = UBig;
+
+    #[inline]
+    fn gcd(self, rhs: &UBig) -> UBig {
+        match (self.into_repr(), rhs.repr()) {
+            (Small(word0), Small(word1)) => UBig::gcd_word(word0, *word1),
+            (Small(word0), Large(buffer1)) => UBig::gcd_large_word(buffer1, word0),
+            (Large(buffer0), Small(word1)) => UBig::gcd_large_word(&buffer0, *word1),
+            (Large(buffer0), Large(buffer1)) => UBig::gcd_large(buffer0, buffer1.clone()),
+        }
+    }
+}
+
+impl Gcd<UBig> for &UBig {
+    type Output = UBig;
+
+    #[inline]
+    fn gcd(self, rhs: UBig) -> UBig {
+        rhs.gcd(self)
+    }
+}
+
+impl Gcd<&UBig> for &UBig {
+    type Output = UBig;
+
+    #[inline]
+    fn gcd(self, rhs: &UBig) -> UBig {
+        match (self.repr(), rhs.repr()) {
+            (Small(word0), Small(word1)) => UBig::gcd_word(*word0, *word1),
+            (Small(word0), Large(buffer1)) => UBig::gcd_large_word(buffer1, *word0),
+            (Large(buffer0), Small(word1)) => UBig::gcd_large_word(buffer0, *word1),
+            (Large(buffer0), Large(buffer1)) => UBig::gcd_large(buffer0.clone(), buffer1.clone()),
+        }
+    }
+}
+
+impl ExtendedGcd<UBig> for UBig {
```

Comment:
Same as `Gcd`: I don't think a trait and overloads are worth it, because any performance benefits of taking by value are slim.

Therefore, make this an inherent function of `UBig` rather than a trait, and always take arguments by reference.

Reply:
Done